File: instrumentation/frontend/Tracing.go
// instrumentation/frontend/tracing.go
//
// REFERENCE IMPLEMENTATION — demonstrates the correct OpenTelemetry SDK
// wiring pattern for a Go service. This code was NOT compiled into the
// running application; the traces actually verified in Kibana came from
// otel-demo's own pre-built, SDK-instrumented images (see
// docs/DECISIONS.md ADR-009). This file shows what full manual/custom
// instrumentation looks like for the pattern the assessment specifies.

package tracing

import (
	"context"
	"net/http"
	"os"

	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/metric"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
	"go.opentelemetry.io/otel/trace"
	"google.golang.org/grpc"
)

var (
	tracer          = otel.Tracer("frontend")
	meter           = otel.Meter("frontend")
	cartViewCounter metric.Int64Counter
)

// InitTracing configures the SDK to export to the local node's collector
// via the Kubernetes downward API host IP.
func InitTracing(ctx context.Context) (func(context.Context) error, error) {
	agentEndpoint := os.Getenv("NODE_IP") + ":4317"

	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceName("frontend"),
			semconv.ServiceVersion(os.Getenv("APP_VERSION")),
			semconv.DeploymentEnvironment(os.Getenv("DEPLOY_ENV")),
			attribute.String("service.language", "go"),
		),
	)
	if err != nil {
		return nil, err
	}

	traceExp, err := otlptracegrpc.New(ctx,
		otlptracegrpc.WithEndpoint(agentEndpoint),
		otlptracegrpc.WithInsecure(),
	)
	if err != nil {
		return nil, err
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(traceExp),
		sdktrace.WithResource(res),
		sdktrace.WithSampler(sdktrace.AlwaysSample()),
	)
	otel.SetTracerProvider(tp)

	metricExp, err := otlpmetricgrpc.New(ctx,
		otlpmetricgrpc.WithEndpoint(agentEndpoint),
		otlpmetricgrpc.WithInsecure(),
	)
	if err != nil {
		return nil, err
	}
	mp := sdkmetric.NewMeterProvider(
		sdkmetric.WithReader(sdkmetric.NewPeriodicReader(metricExp)),
		sdkmetric.WithResource(res),
	)
	otel.SetMeterProvider(mp)

	cartViewCounter, _ = meter.Int64Counter(
		"frontend.cart.view_count",
		metric.WithDescription("Number of times a user viewed their cart"),
	)

	return func(ctx context.Context) error {
		_ = tp.Shutdown(ctx)
		_ = mp.Shutdown(ctx)
		return nil
	}, nil
}

// HTTPMiddleware wraps every route with otelhttp for auto server spans.
func HTTPMiddleware(handler http.Handler, routeName string) http.Handler {
	return otelhttp.NewHandler(handler, routeName)
}

// GRPCDialOptions ensures downstream gRPC calls carry trace context.
func GRPCDialOptions() []grpc.DialOption {
	return []grpc.DialOption{
		grpc.WithStatsHandler(otelgrpc.NewClientHandler()),
	}
}

// RenderCartTemplate — custom business span #1.
func RenderCartTemplate(ctx context.Context, userID string, itemCount int) (context.Context, trace.Span) {
	ctx, span := tracer.Start(ctx, "render-cart-template",
		trace.WithAttributes(
			attribute.String("user.id", userID),
			attribute.Int("product.count", itemCount),
		),
	)
	cartViewCounter.Add(ctx, 1)
	return ctx, span
}

// ValidateCheckoutForm — custom business span #2.
func ValidateCheckoutForm(ctx context.Context, orderTotal float64) (context.Context, trace.Span) {
	return tracer.Start(ctx, "validate-checkout-form",
		trace.WithAttributes(attribute.Float64("order.total", orderTotal)),
	)
}

File: instrumentation/paymentservice/Tracing.js
// instrumentation/paymentservice/tracing.js
//
// REFERENCE IMPLEMENTATION — see the note at the top of
// instrumentation/frontend/tracing.go. This code demonstrates the
// correct Node.js OTel SDK wiring pattern but was not compiled into the
// running application; verified traces came from otel-demo's own images.

const { NodeSDK } = require('@opentelemetry/sdk-node');
const { getNodeAutoInstrumentations } = require('@opentelemetry/auto-instrumentations-node');
const { OTLPTraceExporter } = require('@opentelemetry/exporter-trace-otlp-grpc');
const { OTLPMetricExporter } = require('@opentelemetry/exporter-metrics-otlp-grpc');
const { PeriodicExportingMetricReader } = require('@opentelemetry/sdk-metrics');
const { resourceFromAttributes } = require('@opentelemetry/resources');
const {
  ATTR_SERVICE_NAME,
  ATTR_SERVICE_VERSION,
} = require('@opentelemetry/semantic-conventions');
const { trace, SpanStatusCode, metrics } = require('@opentelemetry/api');

const agentEndpoint = `http://${process.env.NODE_IP}:4317`;

const sdk = new NodeSDK({
  resource: resourceFromAttributes({
    [ATTR_SERVICE_NAME]: 'paymentservice',
    [ATTR_SERVICE_VERSION]: process.env.APP_VERSION || '1.0.0',
    'deployment.environment': process.env.DEPLOY_ENV || 'assessment',
    'service.language': 'nodejs',
  }),
  traceExporter: new OTLPTraceExporter({ url: agentEndpoint }),
  metricReader: new PeriodicExportingMetricReader({
    exporter: new OTLPMetricExporter({ url: agentEndpoint }),
  }),
  instrumentations: [getNodeAutoInstrumentations()],
});

sdk.start();

const tracer = trace.getTracer('paymentservice');
const meter = metrics.getMeter('paymentservice');

const paymentAttempts = meter.createCounter('paymentservice.charge.attempts', {
  description: 'Number of charge attempts, labeled by outcome',
});

function validatePaymentInfo(cardInfo, orderTotal) {
  return tracer.startActiveSpan('validate-payment-info', (span) => {
    span.setAttribute('order.total', orderTotal);
    span.setAttribute('card.type', cardInfo.cardType || 'unknown');
    try {
      if (!cardInfo.number || cardInfo.number.length < 12) {
        throw new Error('invalid card number');
      }
      span.setStatus({ code: SpanStatusCode.OK });
      return true;
    } catch (err) {
      span.recordException(err);
      span.setStatus({ code: SpanStatusCode.ERROR, message: err.message });
      paymentAttempts.add(1, { outcome: 'validation_failed' });
      throw err;
    } finally {
      span.end();
    }
  });
}

async function chargeCard(chargeFn, cardInfo, amount, userId) {
  return tracer.startActiveSpan('charge-card', async (span) => {
    span.setAttribute('user.id', userId);
    span.setAttribute('order.total', amount);
    try {
      const result = await chargeFn(cardInfo, amount);
      span.setAttribute('transaction.id', result.transactionId);
      span.setStatus({ code: SpanStatusCode.OK });
      paymentAttempts.add(1, { outcome: 'success' });
      return result;
    } catch (err) {
      span.recordException(err);
      span.setStatus({ code: SpanStatusCode.ERROR, message: err.message });
      paymentAttempts.add(1, { outcome: 'charge_failed' });
      throw err;
    } finally {
      span.end();
    }
  });
}

module.exports = { sdk, validatePaymentInfo, chargeCard };